Make libgrate usable externally and some fixes along with it

digetx opened the pull request:
Now libgrate.so and related headers will be installed by "make install". I'm using it for experimental opentegra EXA accel, works good so far. So I added some functionality required for EXA (like texture blit) that is not used by grate itself, maybe it will be useful later.

digetx committed:
grate: Makefile: prepend lex/yacc sources with $srcdir

This fixes cross-compilation by Gentoo's crossdev.

digetx committed:
install libgrate.so and relevant headers

digetx committed:
add missed header files to Makefile.am

digetx committed:
add .pc file for libgrate

digetx committed:
grate: add "nodisplay" grate init option

Display stuff won't be created when it's not needed.

digetx committed:
libhost1x: partially convert HOST1X() macros to inlined functions

This fixes -Wshadow compiler warning.

digetx committed:
libhost1x: add host1x_gr2d_clear_rect()

It takes position and size of a rect to be cleared.

digetx committed:
grate: add grate_texture_clear_rect()

It uses host1x_gr2d_clear_rect().

digetx committed:
grate: add grate_init_with_fd()

It takes DRM FD as an argument and passes to host1x_drm_open() for use.

digetx committed:
grate: add grate_texture_blit()

It performs copy of source texture area to destination texture with a resize if necessary.

digetx committed:
libhost1x: handle copy directions in host1x_gr2d_blit()

digetx committed:
libhost1x: handle tileness in host1x_gr2d_clear_rect()

digetx committed:
grate: add grate_bo_create_and_map() helper

The mapping is optional and skipped if map argument is NULL.

digetx committed:
grate: fragment disasm: fix swapped TEX/TXB instructions

digetx committed:
static linking of tools and host1x-grate tests

I run in a problem with a wrong path being resolved by a dynamic library loader inside
bind-mounted directory over NFS. Don't know how to fix it correctly, static linking
is a workaround.

digetx committed:
libhost1x: support Y flip in host1x_gr2d[_surface]_blit

Enable Y flip if destination height argument is negative.

digetx committed:
libhost1x: flip x11-display output

The image produced by GR3D is flipped vertically, now GR2D can unflip it.

kusma commented on include/host1x.h:
```diff
@@ -161,12 +161,12 @@ int host1x_overlay_set(struct host1x_overlay *overlay,
 			__FILE__, __LINE__);				\
-	bo;								\
+	_bo_;								\
 })
```
Hmpf, this is a bit nasty IMO (having to worry about outer scopes just to avoid using an inline function, underscore identifiers being reserved by the compiler)... Maybe sticking the logic inside a static-inline helper is nicer? Something like

```c
static inline struct host1x_bo *host1x_bo_create_helper(struct host1x *host1x, size_t size, int flags, const char *file, int line)
{
	struct host1x_bo *bo = host1x_bo_create(host1x, size, flags);
	if (!bo)
		fprintf(stderr,
 			"ERROR: %s:%d: host1x_bo_create() failed\n",
			file, line);
	return bo;
}
#define HOST1X_BO_CREATE(host1x, size, flags) host1x_bo_create_helper(host1x, size, flags, __FILE__, __LINE__)
```
instead?

digetx replied:
I don't mind, both clunky to me :)

kusma replied:
TBH, I don't really care much for those macros myself. All they do is add some printing, they don't prevent the calling code from misbehaving as a result...

digetx replied:
I don't like when something fails silently.

kusma replied:
Well, maybe we should add a call to abort here in that case?

digetx replied:
Won't hurt, probably.

digetx replied:
However it will hurt in case of opentegra use, as it may fail easily on OOM. CMA quickly gets fragmented.

kusma replied:
Yeah, but then either the caller will have to deal with this (and in that case, it should probably do the error-reporting as well), or else we'll probably crash soon anyway

digetx replied:
In case of opentegra, it fallbacks to non-acceleratable allocation. I don't mind grate tests crashing, it's likely to be something much more severe than OOM.

kusma replied:
Can't opentegra call the non-wrapped version of host1x_bo_create, then?

kusma replied:
Just to be clear, these wrappers are not important to me, do as you please with them. If you want to keep this patch as-is, I'm fine with that. This was just a suggestion.

digetx replied:
Opentegra calls grate_create_texture(), which calls the macro.

I already re-worked it based on your comments ;)

kusma replied:
Right, and grate_create_texture already deals with this failing. Arguably, it wouldn't need to, it could call the un-wrapped version and print itself. But whatever ;)

digetx replied:
I think the state of current error-handling in grate is good enough, at least you'll now if something gone wrong and where to start to look at. Grate is mostly a dev tool, graceful error-handling and avoiding of a tiny memleaks is always good to have but not necessary here IMO.

kusma replied:
Sure, I agree.

kusma commented on include/host1x.h:
```diff
@@ -370,19 +370,24 @@ struct host1x_gr3d;
+		       struct host1x_pixelbuffer *pixbuf,
+		       uint32_t color,
+		       unsigned int x, unsigned int y,
+		       unsigned int width, unsigned int height);
```
maybe call this one `host1x_gr2d_clear_rect` instead?

kusma commented on src/libgrate/grate-texture.c:
```diff
@@ -274,6 +274,18 @@ void grate_texture_clear(struct grate *grate, struct grate_texture *tex,
 
+void grate_texture_clear2(struct grate *grate, struct grate_texture *tex,
+			 uint32_t color, unsigned x, unsigned y,
+			 unsigned width, unsigned height)
```
...and this one `grate_texture_clear_rect`?

digetx replied:
Probably we should rename grate_texture to grate_surface.

kusma replied:
@digetx: Yeah, probably. But that feels like a separate change... Do as you like ;)

kusma commented on src/libgrate/grate.c:
```diff
@@ -113,22 +135,25 @@ bool grate_parse_command_line(struct grate_options *options, int argc,
 }
 
-struct grate *grate_init(struct grate_options *options)
+struct grate *grate_init2(struct grate_options *options, int fd)
```
perhaps `grate_init_fd` (or `grate_init_with_fd`)?

kusma commented on src/libhost1x/host1x-gr2d.c:
```diff
@@ -269,19 +299,130 @@ int host1x_gr2d_clear(struct host1x_gr2d *gr2d,
+	p1_bpp          = PIX_BUF_FORMAT_BYTES(src->format);
+	p1_pitch        = src->pitch;
+	p1_top_left     = src->bo->offset + p1_pitch * sy + sx * p1_bpp;
+	p1_bottom_right = p1_top_left + p1_pitch * height;
```
I think this function is needlessly complicated, because it seems to allow interpreting overlapping memory as wildly different things. Sure, sub-allocations will allow multiple allocations of different format/pitches inside the same BO. But those allocations will not overlap. There's some use-cases where untyped textures are used (but not on T20-class hardware, more like OpenGL 4.x hardware), but in these cases the texel-size is known (you can overlay float32 on int32 and rgba8888 etc, but not on int16 etc). Two separate sub-allocations shouldn't overlap in memory.

So, I think we can first early-out on `src_orig != dst_orig`, then on `src_orig->offset != dst_orig->offset` (different sub-allocations). Then we can assert that the sizes are the same (just in case we somehow botched a sub-allocation). After that, the rest becomes a lot easier, as we *know* that the format and pitch have to be the same. Basially, we need to do a rectangle overlap test, and finally compare the top-left points.

digetx replied:
Yeah, we probably won't ever need such unusual sub-allocations. 👍

kusma replied:
Yeah, just assert all the assumptions in case we ever change our minds ;)

kusma reviewed (changes requested):
Generally very nice, but I have a few nit-picks here ;)

digetx committed:
remove *_pushbuf_create() definitions

These functions do not exist.

digetx committed:
add "needs_explicit_vsync" display parameter for x11-display

The x11-display doesn't display single-buffered grate tests as it requires explicit
framebuffer flip. Grate will perform the flip if vsync option is set (disabled by default),
now it can check whether explicit vsync is needed and force-enable the vsync option.

digetx commented:
Updated the patchset, please take a look.

kusma reviewed (approved):
looks good to me!

digetx commented:
Awesome! 😺
